### PR TITLE
Trim down the Travis build

### PR DESCRIPTION
**Pull request by dtolnay**
This takes us from 28 minutes to 5 minutes. The biggest savings is not compiling all of Clippy + Syntex in release mode for the benchmark.....

### CODE REVIEW

**Comment by dtolnay** [comment]
Related: https://github.com/rust-lang/cargo/issues/2234 Cargo builds build-dependencies (syntex) in release mode when building benchmarks.

File: .travis.yml [hunk 1]
```diff
@@ -20,27 +25,20 @@ before_script:
...
-  popd
 - |
-    (cd json && travis-cargo build) &&
-    (cd json && travis-cargo test) &&
-    (cd json_tests && travis-cargo test -- --features with-syntex --no-default-features) &&
-    (cd json_tests && travis-cargo --only nightly test -- features nightly-testing) &&
-    (cd json_tests && travis-cargo --only nightly bench -- features nightly-testing) &&
-    (cd json_tests && travis-cargo --only nightly bench -- --features "nightly-testing with-syntex" --no-default-features) &&
-    (cd json && travis-cargo --only stable doc)
+    if [ -z $BENCH ]; then
```
Comment (oli-obk) [inline .travis.yml]: I don't epeak bash... What does the -z do, and why is the bench flag set to 1 on nightly?

**Comment by oli-obk** [comment]
Is there any point in running benchmarks on travis?

File: .travis.yml (hunk 1 above)
Comment (dtolnay) [inline .travis.yml]: `-z` means empty i.e. `BENCH.len() == 0`. We run four builds: the three release channels stable beta and nightly, and a fourth one that is nightly with an environment variable BENCH=1. The bench build was the longest so I split it out into it's own build to increase parallelism. The stable/beta/nightly builds all run the first part of this `if` statement. The BENCH=1 build runs the second part which is just `travis-cargo bench`.

File: .travis.yml [hunk 2]
```diff
@@ -7,6 +7,11 @@ rust:
 - beta
 - nightly
```
Comment (dtolnay) [inline .travis.yml]: This is where the first three builds are declared.

File: .travis.yml [hunk 3]
```diff
@@ -7,6 +7,11 @@ rust:
...
 
+matrix:
+  include:
```
Comment (dtolnay) [inline .travis.yml]: This is where the fourth build is declared.

**Comment by dtolnay** [comment]
> Is there any point in running benchmarks on travis?

Basically to make sure it compiles. Also I guess to have a record of performance to look back on when trying to identify what commit introduced a performance regression - it is faster than running `cargo bench` locally. The way I have it, it adds no time to the Travis build because it runs in parallel with the other three builds.

**Comment by oli-obk** [comment]
Iirc cargo test also compiles the benches. But having the travis logs for future reference is a good idea. Thanks for the detailed explanation

@homu r+

**Pull request merged**